internal/raw/rpc_client.go

```diff
@@ -10,22 +15,99 @@ import (
+ * ⠀⣸⣿⣿⣿⣿⣿⣿⣿⣿⠿⠁⠀⡬⣿⣿⣿⣿⣷⣿⣿⣿⠋⠀⢚⣝⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣯⣎⠈⢈⠈⠀⠀⠱⣿⣿⣿⣿⣿⣰⠀⣀⠩⠄⠀⠀⠀⠀⠀⠀⠀⠀
+ * ⣀⣿⣿⣿⣿⣿⣷⣿⣿⠿⠀⢀⠶⣬⣿⣿⡷⠓⣾⣿⣿⣏⣬⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣯⢪⠀⠀⠐⣷⣿⣿⣿⠞⠈⣼⣿⠌⠀⠀⠀⠀⠀⠀⠀⠀
+ * ⣸⣿⣿⣿⡿⢁⣿⣿⠓⠀⣀⢀⣼⣩⣿⣿⢌⣼⣻⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠄⠀⠀⠀⠐⣿⣿⣿⣿⠌⠷⣿⠏⠀⠀⠀⠀⠀⠀⠀⠀
+ * ⣿⣿⣿⣿⣯⣿⣿⢏⠀⠀⠀⠳⢓⣿⣿⠟⣹⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣮⣌⠈⠀⡰⣿⣿⣿⣯⠃⡰⠇⠀⠀⠀⠀⠀⠀⠀⠀
+ * ⣿⣿⣿⣿⣿⣿⣿⡿⠳⢀⠜⣳⣿⣿⡳⣬⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠳⣿⣿⣿⣿⣿⣿⣿⢿⠃⠀⠀⣳⣿⣿⡿⣬⠀⣷⣮⢯⠈⠀⠀⠀⠀⠀
+ * ⣿⣿⣿⣿⣿⣿⣿⠃⢐⣿⠷⣿⡿⣉⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡿⠿⠓⠓⠓⠀⠀⠰⠑⣿⠓⣿⣿⠿⣳⠁⠀⠀⣰⣿⣿⣃⣿⠀⡰⣷⣿⣱⠀⠀⠀⠀⠀
+ * ⣿⣿⣿⣿⣿⡿⠁⠀⠀⠑⣿⢗⣸⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡿⡳⠀⠀⠀⠀⠈⣀⢌⠀⠀⠀⠁⡰⣷⣿⠃⠀⢀⠀⠀⣰⣿⣿⣿⣿⠀⠀⣼⣿⡿⠀⠀⠀⠀⠀
+ * ⣿⣿⣿⣿⣿⠃⡀⠀⢀⢎⠀⡸⣷⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣯⣌⣌⣎⢌⢈⠈⠀⠀⢘⠀⠀⠀⠀⠀⠀⠀⠀⠀⣼⠀⠀⣾⣿⣿⣿⣿⢌⣾⣿⣿⠏⠀⠀⠀⠀⠀
+ * ⣿⣿⣿⣿⡿⢀⠌⣠⡿⠁⠐⣡⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣝⣙⠝⣽⢟⢝⢀⠁⠑⡣⡦⠓⠀⠀⠀⠀⠀⠀⠀⢀⣼⠏⠀⣨⣿⣿⣿⣿⣿⣿⣿⣿⣿⣃⠀⠀⠀⠀⠀
+ * ⣷⣿⣿⣿⠏⣼⠀⣸⢇⠀⣈⣷⡿⣳⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡹⣓⣟⠐⠀⠀⠀⢀⣮⣿⣿⣿⣿⣾⣬⢟⠈⠁⠀⣿⣿⣿⣿⣿⣿⠻⣿⣿⣿⢟⠌⠀⡠⣏⠀
+ * ⣾⣿⣿⣿⠏⠏⠌⣿⠞⣨⠗⣼⠉⠓⣼⣷⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣯⣾⣿⣎⠈⠀⣈⣿⣿⣿⣿⣿⣷⣿⣿⣿⣿⠀⠀⣿⣿⣿⠟⠟⣕⠁⠐⡳⣿⣿⠋⣈⣮⠌⠀
+ * ⣿⣿⣿⣿⣿⠏⣯⣿⣫⣿⠾⠀⠀⠘⡻⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡿⠷⢳⠳⡳⣿⣿⣿⣿⣯⣿⣿⣿⣿⠷⠇⠀⡰⠰⣿⣿⣿⠆⠀⣿⣿⣿⣿⣿⣾⠎⠀⠀⣱⣿⣿⣿⣿⠏⠀
+ * ⣰⣿⣿⣿⣷⠇⡿⣷⣿⣿⣧⡧⡦⣷⣿⣿⣿⣿⣿⣿⣿⣿⣿⡿⣿⣿⣿⣿⣿⠣⠁⠀⠀⠀⠀⣷⣿⣿⣿⣿⣿⣿⡿⠇⠀⠀⠀⠀⠀⣰⣿⣿⠎⠀⣿⣿⣿⣿⣿⣿⣯⠀⠀⠀⣿⣿⣿⣿⠂⠀
+ * ⣰⣿⣿⣿⠄⣨⣯⣾⣿⣿⠀⠀⢀⡾⣹⣿⣿⣿⣿⣿⣿⣿⢓⣶⣿⣿⣿⣿⠝⠀⠀⠀⠀⠀⢀⣠⣿⣿⣿⣷⣿⣿⠏⠀⠀⠀⠀⠀⠀⠀⣿⣿⣿⠈⡱⣿⣿⣿⣿⣿⠟⣎⠀⠀⣿⣿⣿⣿⠏⠀
+ * ⠾⣿⣿⣿⠎⣿⣿⣿⣿⢟⡜⠆⠸⠱⣿⣿⣿⣿⣿⣿⡿⠁⠀⣿⣿⣿⣿⣿⠀⠀⠀⠀⠀⠀⠀⣸⣿⠷⠁⠒⠑⣿⠏⠀⠀⠀⠀⠀⠀⣰⣿⣿⣿⠁⠀⣳⣿⣿⣿⣿⣏⣷⠎⠀⣿⣿⣿⣿⠏⠀
+ * ⣿⣿⣿⣿⠌⣱⣿⣿⣿⣿⠏⢀⣦⣾⣿⣿⣿⣿⣿⠿⣀⠀⢎⣼⣿⣿⣿⣿⠀⠀⠀⠀⠀⡈⠀⣿⣿⠗⠀⠀⠀⣰⣿⠈⠀⠀⠀⠀⠠⣿⣿⣿⠗⠀⠀⠰⣿⣿⣿⣿⣿⣾⠏⣰⣿⣿⠉⠐⠀⡏
+ * ⣿⣿⣿⣿⠉⣰⣿⣿⣿⣿⣿⣾⣿⣿⣼⣿⣿⣿⣿⢃⠀⠀⣿⣻⣿⣿⣿⣿⣈⠈⠀⠀⠀⠀⣼⣿⣿⣿⢎⠀⣬⠐⠀⠁⣀⢌⠀⣀⣿⣿⣿⠷⠀⠀⠀⠀⣳⣿⣿⣿⣷⡿⣫⣿⣿⣿⠏⠀⠀⡀
+ * ⣿⣿⣿⣿⠏⣰⣿⣿⣿⣿⣿⣿⣷⣿⣿⣿⣿⣿⠏⠀⠀⡀⣫⣿⣿⣿⣿⣿⣿⣿⣿⣼⣯⣿⣿⣿⣿⣿⣿⣾⠆⠀⠀⠀⠐⣿⣿⣿⣿⣿⢷⠀⠀⠀⠀⠀⣸⣿⣿⠟⠐⣿⣿⣿⣿⣿⠏⠀⠀⠀
+ * ⣿⣿⣿⣿⣿⠐⢏⠏⣿⣿⣿⣿⠰⡿⣿⣷⣿⣿⠁⠀⢀⠀⣳⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣳⣿⣿⠿⠀⠀⠀⠀⠀⣿⣿⣿⣿⡟⠀⠀⠀⠀⠀⣀⣿⣿⣿⠀⠀⣰⣿⡿⣷⣿⠃⠀⠀⠀
+ * ⣿⣿⣿⣿⣿⣠⣾⠁⠏⠗⣿⠏⠀⠀⢏⣰⠟⠐⠏⠀⠀⠀⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠇⣿⣿⣿⠏⠀⠀⠀⠀⣠⣿⣿⡿⠑⠀⠀⠀⠀⠀⢀⣾⣿⣿⠿⠀⠀⣰⣿⣿⣆⠐⢷⠀⠀⠀
+ * ⣳⣿⣿⣿⣿⠰⣿⢌⠑⠆⠟⣳⠀⠀⠑⠐⠁⠀⢀⠀⠀⠀⠱⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡿⠓⣰⣿⣿⣿⣏⣌⢌⠀⣠⣿⣿⠿⠁⠀⠀⠀⠀⠀⣈⣾⣿⣿⣿⢟⠈⠀⣼⣿⣿⠀⠀⠀⠀⠀⠀
+ * ⣾⣿⣿⣿⣿⠏⣿⣿⠏⡦⠀⠱⠀⠀⣰⠈⡀⠀⣸⠌⠀⡀⠀⣿⣳⣿⣿⣿⣿⣿⣿⠿⡻⠁⣨⣿⣿⣿⣿⣿⣿⣿⡌⠌⣰⣿⠃⠀⠀⠀⠀⣌⣾⣿⣿⣿⣿⣿⣷⠀⣀⣿⣿⣿⠃⠀⠀⠀⠀⠀
+ * ⣿⣿⣿⣿⣿⣿⣿⣿⣿⣯⡈⠀⠄⠀⠠⣀⠀⡤⣿⣯⠌⠠⠠⠀⠒⡷⡷⣿⡿⠳⠀⠀⢀⣬⣿⣿⣿⣿⣿⣿⣿⣿⣯⢌⣰⢟⠀⠀⠀⠀⢀⣿⣿⣿⣿⣿⣿⠗⠂⣨⣿⣿⣿⣿⠏⠀⠀⠀⠀⠀
+ * ⣿⣿⣿⡿⣿⣿⣿⣿⣿⣿⣿⣯⢌⢌⢈⣨⣨⢀⣿⣿⣿⣮⣬⠌⠎⠀⠀⠀⠀⠀⢀⣬⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣯⣎⠀⠀⠀⣼⣿⣿⣿⣿⠗⡿⢀⣾⣿⠿⣿⣿⡿⠁⠀⠀⠀⠀⠀
+ * ⣿⣿⣿⠇⡰⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠏⣼⣿⣿⣿⣿⣿⣿⣍⠀⠈⣌⢀⠌⣾⢓⠗⣿⣿⣿⣿⣿⣿⣿⣿⡷⣷⠷⣿⡿⠁⠀⠀⣼⣿⣿⣿⡽⢃⢸⣭⣿⣿⣿⣏⣱⣿⢉⠎⠲⠀⠀⠀⠀
+ * ⣷⣿⣟⠀⢀⣰⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠀⣿⣿⣿⣿⡳⣿⣿⣿⠃⠁⠑⡰⣷⣿⣾⠀⠐⠀⣱⣿⠿⠁⡾⠷⠀⠀⡠⠓⠀⠀⢈⠀⡱⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠄⣿⣿⠏⠀⠀⠀⠀⠀
+ * ⣼⣿⣿⠀⠀⣰⣿⣿⣿⣿⣿⣿⡿⣿⣿⣿⠀⠷⣽⣿⣿⢎⠃⣱⣿⢎⢈⣈⢈⠀⠀⠑⠀⠀⢀⠀⢓⢀⠀⢈⠈⠀⢀⢈⠀⠐⣯⣿⠏⠀⣷⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠀⣼⣿⠏⠂⠀⠀⠀⠀
+ * ⣿⣿⣿⣏⠈⡰⣿⣿⣿⣿⣿⣿⠀⣿⣿⡿⣰⣀⠿⣿⣿⣿⣏⠘⣳⣿⣿⣿⣿⣮⣜⣬⣮⣞⣿⣯⣿⣿⣏⣿⣿⣿⣾⣿⣿⣿⣿⣿⣯⠀⣰⣿⣿⣿⣿⣿⣿⣿⣿⠗⠃⠀⣿⣿⠏⠀⠀⠀⠀⠀
+ * ⣱⣿⣿⣿⣿⣮⠽⡳⣿⣿⣿⠏⠀⣿⣿⣯⢘⣰⣁⣇⠿⣼⣿⣿⠞⠱⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠀⣰⣿⣿⣿⣿⠀⡰⠓⠑⠀⠀⣼⣿⣿⠃⠀⠀⠀⠀⠀
+ * ⢰⣿⣿⣿⣿⣿⠏⣳⣿⣿⣿⣿⠀⣷⣿⣿⣿⠀⠸⣰⠀⢟⣱⣿⣿⣎⣸⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡿⠀⣿⣿⣿⣿⠏⠀⠀⠀⠀⡌⠀⣰⣿⠃⠀⠀⠀⠀⠀⠀
+ * ⣰⣿⣿⡿⡾⣷⣯⣰⣿⣿⣿⣿⣎⣸⣿⣿⣿⣯⠏⣰⠀⠏⣼⣿⣷⣷⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡿⡷⠓⠀⣀⣿⣿⣿⠗⠀⠀⣈⠮⠳⠉⠀⣼⠃⠀⠀⠀⠀⠀⠀⠀
+ * ⣰⣿⣿⠎⠀⡰⣿⣟⣷⣿⣿⣿⣿⠟⡰⣳⣿⣿⣿⣾⣈⠏⣳⠳⡰⠰⡷⣳⠗⠷⡷⣿⣿⣿⣿⣿⣿⣿⠷⡷⡷⠳⠳⠳⠑⠁⠀⠀⣨⣿⣿⣿⠏⠂⣠⡴⠓⠀⠀⠀⠴⠁⠀⠀⠀⠀⠀⠀
+ * TODO: delete this function when this fix reaches stable version of tendermint⠀⠀
+ */
+func setupProxy(targetAddr string, logger *zap.Logger) (string, error) {
+	u, err := url.Parse(targetAddr)
+	if err != nil {
+		return "", fmt.Errorf("%s is not a valid url: %w", targetAddr, err)
+	}
+
+	if u.Scheme == "tcp" {
+		u.Scheme = "http"
+	}
+	proxy := httputil.NewSingleHostReverseProxy(u)
+	originalDirector := proxy.Director
+	proxy.Director = func(request *http.Request) {
+		originalDirector(request)
+		request.Host = u.Host
```

Comment:
the only question is why do you change Host after the original director

Reply:
From https://pkg.go.dev/net/http/httputil@master#NewSingleHostReverseProxy:

>NewSingleHostReverseProxy does not rewrite the Host header.

>To customize the ReverseProxy behavior beyond what NewSingleHostReverseProxy provides, use ReverseProxy directly with a Rewrite function.